### ISSUE TITLE
Memory leak in 2.0.0.dev-Unknown (migrated from Trac #1917)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3465]
Original ticket http://projects.scipy.org/numpy/ticket/1917
Reported 2011-07-28 by trac user jiffyclub, assigned to unknown.

The following code overruns my system's memory in a fraction of a second:

```
>>> import numpy as np
>>> np.__version__
'2.0.0.dev-Unknown'
>>> def test():
...     a = np.zeros((10000,10000),dtype=np.float)
... 
>>> while True:
...     test()
... 
```

I believe this is the most recent numpy, checked out and compiled last night (7-28-2011). I'm using Python 2.7.1 on Mac OS 10.6.8.

**Comment by thouis** [comment]
Comment in Trac by trac user jiffyclub, 2011-07-29

**Comment by thouis** [comment]
Comment in Trac by trac user jiffyclub, 2011-07-29

If it's possible it might be good to add a test for escalating memory usage to the unit tests so this kind of bug can be noticed in the tests.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-07-29

It's not a memory leak, it's your OS not releasing memory.  It's an annoying problem, but can't be fixed in numpy.  See for example http://thread.gmane.org/gmane.comp.python.numeric.general/44009  Also see the tcmalloc link for a solution.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2011-07-29

It' hard to tell exactly when the problem was introduced as all commits don't build, but it was sometime after bff2e5159bb65f7eb939896d0b50a6162f6fd286,

Ralf, it _is_ a memory leak.

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2011-07-29

I can see it very clearly with the following procedure:
1. Open Task Manager/System Monitor to see memory usage
2. Run the following, which shows a jump upwards in memory usage.
   
   > > > import numpy as np
   > > > a = np.zeros((10000,10000),dtype=np.float)
3. Run the following, which drops the memory usage back down in an older version of NumPy, but not in master.
   >>> del a

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-07-29

You know, I was arguing something similar in the thread I linked or another one around the same date.  Doing the exact same thing.  Even got my system to slow down so much it became unresponsive.  Still not a memory leak apparently.  I'm not saying it isn't a leak, but this is not a proof.  What does Valgrind say?

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2011-07-29

Fails for db90bf507279dffa23e6546414b82dd28ecaf984, works for bff2e5159bb65f7eb939896d0b50a6162f6fd286.

Ralf, the fact that the failure is new argues that it was introduced.

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2011-07-29

Found the bug, this was because I forgot to uncomment the PyDataMem_FREE call again after tracking down some difficult crashing issues. I had a double-free issue I tracked down by commenting that out, then in the relief of getting that sorted out, I neglected that important step. I'll try to come up with a test which would catch this kind of thing and commit the fix shortly.
